crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+/// be established between the bridge halves. This link is the argument `corresponding_bridge_half`,
+/// which can both send and receive messages from the other bridge loop.
+///
+/// When a message is forwarded, the [Publish] is forwarded from this loop to the corresponding loop.
+/// This allows the loop to store the message along with its packet ID when the forwarded message is
+/// published. When an acknowledgement is received for the forwarded message, the packet id is used
+/// to retrieve the original [Publish], which is then passed to [AsyncClient::ack] to complete the
+/// final step of the message flow.
+///
+/// The channel sends [Option<Publish>] rather than [Publish] to allow the bridge to send entirely
+/// novel messages, and not just forwarded ones, as attaching packet IDs relies on pairing every
+/// [Outgoing] publish notification with a message sent by the relevant client. So, when a QoS 1
+/// message is forwarded, this will be accompanied by sending `Some(message)` to the channel,
+/// allowing the original message to be acknowledged once an acknowledgement is received for the
+/// forwarded message. When publishing a health message, this will be accompanied by sending `None`
+/// to the channel, telling the bridge to ignore the associated packet ID as this didn't arise from
+/// a forwarded message that itself requires acknowledgement.
+///
+/// # Health topics
+/// The bridge will publish health information to `health_topic` (if supplied) on `target` to enable
+/// other components to establish bridge health. This is intended to be used the half with cloud
+/// event loop, so the status of this connection will be relayed to a relevant `te` topic like its
+/// mosquitto-based predecessor. The payload is either `1` (healthy) or `0` (unhealthy). When the
+/// connection is created, the last-will message is set to send the `0` payload when the connection
+/// is dropped.
+async fn half_bridge<F: for<'a> Fn(&'a str) -> Cow<'a, str>>(
+    mut recv_event_loop: EventLoop,
+    target: AsyncClient,
+    transform_topic: F,
+    mut companion_bridge_half: BidirectionalChannelHalf<Option<Publish>>,
+    health_topic: Option<String>,
+    name: &'static str,
+) {
+    let mut forward_pkid_to_received_msg = HashMap::new();
+    let mut last_err = Some("dummy error".into());
+
+    loop {
+        let notification = match recv_event_loop.poll().await {
+            Ok(notification) => {
+                if last_err.as_ref().is_some() {
+                    info!("MQTT bridge connected to {name} broker");
+                    last_err = None;
+                    if let Some(health_topic) = &health_topic {
+                        target
+                            .publish(health_topic, QoS::AtLeastOnce, true, C8Y_BRIDGE_UP_PAYLOAD)
+                            .await
+                            .unwrap();
+                        companion_bridge_half.send(None).await.unwrap();
```

Comment:
This line deserves a comment, as having a de-synchronization of the two bridge halves is the Achille's heel of this algorithm.